### PR TITLE
Add edition support to rules and toolchain. (#157)

### PR DESCRIPTION
**Pull request by mfarrugi**
Fairly simple change, backwards compatible through optional/default parameters.

`rust_toolchain.default_edition` is not practically useful until cargo-raze is providing edition.

### CODE REVIEW

File: rust/private/rust.bzl [hunk 1]
```diff
@@ -183,6 +183,9 @@ _rust_common_attrs = {
     ),
     "deps": attr.label_list(),
     "crate_features": attr.string_list(),
+    "edition": attr.string(
+        doc = "The rust edition to use for this crate.",
```
Comment (mfarrugi) [inline rust/private/rust.bzl]: This is apparently the way forward, added note in #140

**Comment by mfarrugi** [comment]
Ah forgot, need to update rustc to 1.31 in this repo.. will get to that a little later.

File: rust/private/rustc.bzl [hunk 2]
```diff
@@ -217,6 +217,7 @@ def rustc_compile_action(
     if hasattr(ctx.attr, "crate_features"):
         args.add_all(getattr(ctx.attr, "crate_features"), before_each = "--cfg", format_each = 'feature="%s"')
     args.add_all(rust_flags)
+    args.add("--edition", _get_edition(ctx, toolchain))
```
[the remark this answers is not in the document]
Comment (mfarrugi) [reply rust/private/rustc.bzl]: I'll omit the flag when it is "2015".

Good catch on rustdoc, those invocations are not passing --edition...

**Comment by mfarrugi** [comment]
Should be good now, lmk.

File: rust/private/rustdoc_test.bzl [hunk 3]
```diff
@@ -66,29 +66,34 @@ def _build_rustdoc_test_script(toolchain, dep_info, crate):
...
+    link_flags += ["--extern=" + crate.name + "=" + crate.output.short_path]
+    link_flags += ["--extern=" + c.name + "=" + c.output.short_path for c in d.direct_crates.to_list()]
     link_search_flags += ["-Ldependency={}".format(dirname(c.output.short_path)) for c in d.transitive_crates.to_list()]
 
     link_flags += ["-ldylib=" + get_lib_name(lib) for lib in d.transitive_dylibs.to_list()]
     link_search_flags += ["-Lnative={}".format(dirname(lib.short_path)) for lib in d.transitive_dylibs.to_list()]
     link_flags += ["-lstatic=" + get_lib_name(lib) for lib in d.transitive_staticlibs.to_list()]
     link_search_flags += ["-Lnative={}".format(dirname(lib.short_path)) for lib in d.transitive_staticlibs.to_list()]
 
+    edition_flags = ["--edition={}".format(crate.edition)] if crate.edition != "2015" else []
```
Comment (acmcarther) [inline rust/private/rustdoc_test.bzl]: Really minor nit: Can this be converted to use the add_addition_flag fn?

Comment (mfarrugi) [reply rust/private/rustdoc_test.bzl]: I couldn't think of a way to get the fn to work on both 'args' and 'list

i'll take a look at the rustdoc_test todo to use args

Comment (mfarrugi) [reply rust/private/rustdoc_test.bzl]: filed https://github.com/bazelbuild/bazel/issues/6888 for args issue

**Pull request merged**